static/app/views/dashboards/widgetCard/autoSizedText.tsx

```diff
@@ -0,0 +1,177 @@
+import {useLayoutEffect, useRef, useState, useTransition} from 'react';
+import styled from '@emotion/styled';
+import {useResizeObserver} from '@react-aria/utils';
+import debounce from 'lodash/debounce';
+
+interface Props {
+  children: React.ReactNode;
+  maxFontSize: number;
+  minFontSize: number;
+  calculationCountLimit?: number;
+}
+
+export function AutoSizedText({
+  children,
+  minFontSize,
+  maxFontSize,
+  calculationCountLimit = DEFAULT_CALCULATION_COUNT_LIMIT,
+}: Props) {
+  const [_isPending, startTransition] = useTransition();
+
+  const [parentHeight, setParentHeight] = useState<number | null>(null);
+  const [parentWidth, setParentWidth] = useState<number | null>(null);
+
+  const [fontSize, setFontSize] = useState<number>((maxFontSize + minFontSize) / 2);
+  const [fontSizeLowerBound, setFontSizeLowerBound] = useState<number>(minFontSize);
+  const [fontSizeUpperBound, setFontSizeUpperBound] = useState<number>(maxFontSize);
+
+  const [calculationCount, setCalculationCount] = useState<number>(0);
+
+  const parentRef = useRef<HTMLDivElement>(null);
+  const childRef = useRef<HTMLDivElement>(null);
+
+  const onResize = debounce(() => {
+    const parentDimensions = getElementDimensions(parentRef.current);
+
+    if (parentDimensions) {
+      // Reset the known font size bounds, they're no longer valid
+      setCalculationCount(0);
+      setFontSizeLowerBound(minFontSize);
+      setFontSizeUpperBound(maxFontSize);
+
+      // Update state, trigger a re-render
+      setParentHeight(parentDimensions.height);
+      setParentWidth(parentDimensions.width);
+    }
+  }, RESIZE_DEBOUNCE);
+
+  useResizeObserver({
+    ref: parentRef,
+    onResize,
+  });
+
+  useLayoutEffect(() => {
+    if (calculationCount > calculationCountLimit) {
+      // Exceeded the iteration count. This should be unlikely! If it happens, abandon
+      return;
+    }
```

Comment:
It would be interesting to collect some telemetry here in case this _does_ happen, so we can check out what the conditions are.